File: elixir/space-age/space_age.exs
defmodule SpaceAge do
  @type planet :: :mercury | :venus | :earth | :mars | :jupiter
                | :saturn | :neptune | :uranus

  @doc """
  Return the number of years a person that has lived for 'seconds' seconds is
  aged on 'planet'.
  """
  @spec age_on(planet, pos_integer) :: float
  def age_on(planet, seconds) do
    case {planet} do
      {:earth} -> seconds / 31557600
      {:mercury} -> seconds / 0.2408467 / 31557600
      {:venus} -> seconds / 0.61519726 / 31557600
      {:mars} -> seconds / 1.8808158 / 31557600
      {:jupiter} -> seconds / 11.862615 / 31557600
      {:saturn} -> seconds / 29.447498 / 31557600
      {:neptune} -> seconds / 164.79132 / 31557600
      {:uranus} -> seconds / 84.016846 / 31557600
    end
  end
end

File: haskell/space-age/src/SpaceAge.hs
module SpaceAge (Planet(..), ageOn) where

data Planet = Mercury
            | Venus
            | Earth
            | Mars
            | Jupiter
            | Saturn
            | Uranus
            | Neptune

ageOn :: Planet -> Float -> Float
ageOn Mercury seconds = earthSeconds seconds / 0.2408467
ageOn Venus seconds = earthSeconds seconds / 0.61519726 
ageOn Earth seconds = earthSeconds seconds
ageOn Mars seconds = earthSeconds seconds / 1.8808158 
ageOn Jupiter seconds = earthSeconds seconds / 11.862615
ageOn Saturn seconds = earthSeconds seconds / 29.447498
ageOn Uranus seconds = earthSeconds seconds / 84.016846
ageOn Neptune seconds = earthSeconds seconds / 164.79132 

earthSeconds :: Float -> Float
earthSeconds seconds = seconds / 31557600